contacts/views.py

```diff
@@ -1,230 +1,157 @@
+                serializer.save()
+            return Response(serializer.data, status=HTTP_200_OK)
+        else:
+            return Response("PUT REQUEST FAILED", status=HTTP_400_BAD_REQUEST)
+
+    def get_serialized(self, pk, user):
+
+        queryset = Contact.objects.filter(user_id=pk)
+
+        serializer = ContactSerializer(queryset, many=True)
+        data = serializer.data
+        return data
+
+
+class UserContacts(APIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def get(self, request):
+        pk = request.COOKIES.get('user_id')
+        _id = int(pk)
+        user = User.objects.get(id=_id)
+        contact = Contact.objects.filter(user_id=user.id)
+        serializer = ContactSerializer(contact, many=True)
+        return Response(serializer.data)
+
+    def post(self, request):
+        serializer = ContactSerializer(data=request.data)
+        if serializer.is_valid():
+            serializer.save()
+            return Response(serializer.data, status=status.HTTP_201_CREATED)
+        return Response(serializer.errors)
+
+    def put(self, request):
+        if 'token' in request.COOKIES:
+            contact = Contact.objects.get(id=request.data['id'])
+            payload = request.data
+            serializer = ContactSerializer(contact, data=payload, partial=True)
+            if serializer.is_valid(raise_exception=True):
+                serializer.save()
+            return Response(serializer.data, status=HTTP_200_OK)
+        else:
+            return Response("PUT REQUEST FAILED", status=HTTP_400_BAD_REQUEST)
+
+
+class ContactDetails(APIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def get(self, request, pk):
+        book = Contact.objects.get(id=pk)
+        serializer = ContactSerializer(book)
+        return Response(serializer.data)
+
+
+class UserDetails(APIView):
+    permission_classes = (permissions.AllowAny,)
+
+    def get(self, request, pk):
+        user = User.objects.get(id=pk)
+        contacts = self.get_serialized(pk, user)
+
+        serializer = UserDetailSerializer(user)
+        return Response(serializer.data)
+
+    def get_serialized(self, pk, user):
+
+        queryset = Contact.objects.filter(user_id=pk)
+
+        serializer = ContactSerializer(queryset, many=True)
+        data = serializer.data
+        return data
+
+
+@api_view(["POST"])
+@permission_classes((AllowAny,))
+@csrf_exempt
+def login(request):
+    username = request.GET.get("username")
+    password = request.GET.get("password")
+    if username is None or password is None:
+        return Response({'error': 'Please provide both username and password'},
+                        status=HTTP_400_BAD_REQUEST)
+    user = authenticate(username=username, password=password)
+    if not user:
+        return Response({'error': 'Invalid Credentials'},
+                        status=HTTP_404_NOT_FOUND)
+
+    token, _ = Token.objects.get_or_create(user=user)
```

Comment:
why is the underscore necessary?

Reply:
it gives two things in return and we only concern with the authentication token that's why underscore is needed!